### PR TITLE
[lang] Merged VirtualMemoryAllocator into MemoryPool for LLVM-CPU backend

### PR DESCRIPTION
**Pull request by jim19930609**
Issue: https://github.com/taichi-dev/taichi/issues/7599

The `MemoryPool`, `CpuDevice`, `VirtualMemoryAllocator` and `UnifiedMemoryAllocator` are nested together thus unfortunately this PR is hard to split further. 

### Brief Summary
Key changes happen in `cpu_device.cpp/.h`, `virtual_memory.h` (removed), `memory_pool.cpp/.h`, `unified_allocator.cpp/.h`. All other changes are adaptions to the singleton MemoryPool.

1. `MemoryPool` now becomes a Singleton, instead of a member of `Program`
2. `VirtualMemoryAllocator` is fused into a private method `MemoryPool::allocate_raw_memory()`, which can only get called by its `Allocator` friends. `MemoryPool` is the single owner of the allocated raw memory.
3. `UnifiedAllocator` and `CpuDevice` no longer allocates raw memory on themselves, but indirectly allocates memory via `MemoryPool`
4. All the external memory allocation request will now go through `MemoryPool::allocate()` directly or indirectly (through Device API), so that the `Allocator` inside `MemoryPool` can manage the memory request and release with specific algorithms. 

image (3)

5. In a successive PR, we'll make `Allocators` only accessible by the `MemoryPool` (https://github.com/taichi-dev/taichi/pull/7685)

### CODE REVIEW

**Review by ailzhang — APPROVED** [review]
Thanks!!

File: tests/python/test_ndarray.py [hunk 1]
```diff
@@ -77,12 +76,11 @@ def _test_matrix_ndarray(n, m, dtype, shape):
     assert x.m == m
 
 
-@pytest.mark.parametrize('n,m', matrix_dims)
-@pytest.mark.parametrize('dtype', data_types)
-@pytest.mark.parametrize('shape', ndarray_shapes)
 @test_utils.test(arch=get_host_arch_list())
-def test_matrix_ndarray(n, m, dtype, shape):
-    _test_matrix_ndarray(n, m, dtype, shape)
+def test_matrix_ndarray():
```
[the remark this answers is not in the document]
Comment (jim19930609) [reply tests/python/test_ndarray.py:80]: Ah my bad, thanks!

File: taichi/rhi/cpu/cpu_device.cpp [hunk 2]
```diff
@@ -10,16 +11,24 @@ CpuDevice::AllocInfo CpuDevice::get_alloc_info(const DeviceAllocation handle) {
   return allocations_[handle.alloc_id];
 }
 
+CpuDevice::CpuDevice() {
+#if defined TI_PLATFORM_OSX
+  arch_ = Arch::arm64;
+#else
+  arch_ = Arch::x64;
```
[the remark this answers is not in the document]
Comment (jim19930609) [reply taichi/rhi/cpu/cpu_device.cpp:18]: ah I should have been used `host_arch()` here

[the remark this answers is not in the document]
Comment (jim19930609) [reply taichi/rhi/cpu/cpu_device.cpp:18]: Thanks! Switched to use `host_arch()` here

**Pull request merged**